[grid] Local session request queuer and queue

pujagani opened the pull request:
<!-- NOTE
Please be aware that the Selenium Grid 3.x is being deprecated in favour of the
upcoming version 4.x. We won't be receiving any PRs related to the Grid 3.x code. Thanks!
-->

**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.


<!--- Provide a general summary of your changes in the Title above -->

### Description
Change are for the issue #8558. 
- Adding NewSessionQueue and NewSessionQueuer interfaces. 
- LocalSessionQueue and LocalSessionQueuer are the implementations.
- NewSessionRequest and NewSessionRequestResponse events are added to communicate via the EventBus.
- The incoming request has a request-id which ties the whole flow together. The request-id is used by GetNewSessionResponse to maintain a local map of request and response and return the response as it receives it.

Queuer is responsible for sending the incoming /session request to the appropriate handler and waiting for the response.
LocalSessionQueue will perform the actual operation of adding, removing and retry adding to the local queue.

### Motivation and Context
New session request should not be dropped as it is happening in the current implementation. Instead, it should be queued and handled as the slots are available.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
<!--- Provide a general summary of your changes in the Title above -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/SeleniumHQ/selenium?pullRequest=8689) <br/>All committers have signed the CLA.

shs96c commented on java/server/src/org/openqa/selenium/grid/data/NewSessionErrorResponse.java:
```diff
@@ -0,0 +1,77 @@
+    return requestId;
+  }
+
+  private Map<String, Object> toJson() {
```
Because you have sensibly named getters, you can omit this method.

shs96c commented on java/server/src/org/openqa/selenium/grid/data/NewSessionRequestEvent.java:
```diff
@@ -0,0 +1,34 @@
+
+  public static final Type NEW_SESSION_REQUEST = new Type("new-session-request");
+
+  public NewSessionRequestEvent(UUID requestId)
```
Nit: Java formatting, please

shs96c commented on java/server/src/org/openqa/selenium/grid/data/NewSessionRequest.java:
```diff
@@ -0,0 +1,31 @@
+
+public class NewSessionRequest {
+
+  private final UUID requestId;
```
We may want a `RequestId` class. Passing around `UUID` is getting confusing :)

shs96c commented on java/server/src/org/openqa/selenium/grid/data/NewSessionResponse.java:
```diff
@@ -0,0 +1,91 @@
+  private final byte[] downstreamEncodedResponse;
+  private final UUID requestId;
+
+  public NewSessionResponse(Session session, byte[] downstreamEncodedResponse, UUID requestId) {
```
Nit: the `requestId` is the primary identifier. Feel free to make it the first parameter

shs96c commented on java/server/src/org/openqa/selenium/grid/data/NewSessionResponse.java:
```diff
@@ -0,0 +1,91 @@
+
+  public NewSessionResponse(Session session, byte[] downstreamEncodedResponse, UUID requestId) {
+    this.session = Require.nonNull("Session", session);
+    this.downstreamEncodedResponse = Require
```
Nit: Prefer breaking: 

```
this.downstreamEncodedResponse = Require.nonNull(
    "downstream"....);
```

shs96c commented on java/server/src/org/openqa/selenium/grid/sessionqueue/AddBackToSessionQueue.java:
```diff
@@ -0,0 +1,61 @@
+// specific language governing permissions and limitations
+// under the License.
+
+package org.openqa.selenium.grid.sessionqueue;
```
I think it's safe to call the package `queue` rather than `sessionqueue`, but your call.

shs96c commented on java/server/src/org/openqa/selenium/grid/sessionqueue/AddBackToSessionQueue.java:
```diff
@@ -0,0 +1,61 @@
+
+      boolean value = newSessionQueuer.retryAddToQueue(req, id);
+
+      return new HttpResponse().setContent(
```
Should we add the response to the Span?

pujagani replied:
Sorry misinterpreted this one. Will add it whenever required. Thanks!

shs96c commented on java/server/src/org/openqa/selenium/grid/sessionqueue/GetNewSessionResponse.java:
```diff
@@ -0,0 +1,135 @@
+  private static final Logger LOG = Logger.getLogger(GetNewSessionResponse.class.getName());
+  private final EventBus bus;
+  private final Tracer tracer;
+  public final NewSessionQueue sessionRequests;
```
👀 I doubt we want the queue to be public....

shs96c commented on java/server/src/org/openqa/selenium/grid/sessionqueue/GetNewSessionResponse.java:
```diff
@@ -0,0 +1,135 @@
+    // Each thread will get its own CountDownLatch and it is stored in the Map using request id as the key.
+    // EventBus thread will retrieve the same request and set it's response and unblock waiting request thread.
+    UUID id = sessionResponse.getRequestId();
+    NewSessionRequest sessionRequest = knownRequests.get(id);
```
If `remove` is called before `setResponse` this will cause an NPE.

pujagani replied:
`remove` will be called only after the CountdownLatch countdowns, which happens only when a response is received.

shs96c commented on java/server/src/org/openqa/selenium/grid/sessionqueue/GetNewSessionResponse.java:
```diff
@@ -0,0 +1,135 @@
+  public HttpResponse add(HttpRequest request) {
+    Require.nonNull("New Session request", request);
+
+    CountDownLatch latch = new CountDownLatch(1);
```
Isn't the `NewSessionRequest` effectively a `CompletableFuture`?

pujagani replied:
Yep, it kinda is.  Do you foresee any problems with the CountDownLatch?

shs96c commented on java/server/src/org/openqa/selenium/grid/sessionqueue/GetNewSessionResponse.java:
```diff
@@ -0,0 +1,135 @@
+
+  private void setErrorResponse(NewSessionErrorResponse sessionResponse) {
+    UUID id = sessionResponse.getRequestId();
+    NewSessionRequest sessionRequest = knownRequests.get(id);
```
There's another NPE here.

pujagani replied:
Yep, it is fixed in a later commit. Will push that bit out.

shs96c commented on java/server/src/org/openqa/selenium/grid/sessionqueue/GetNewSessionResponse.java:
```diff
@@ -0,0 +1,135 @@
+}
+
+  private void removeRequest(UUID id) {
+    knownRequests.remove(id);
```
What happens if we're in the middle of waiting for `add` to finish?

pujagani replied:
Good point. We shouldn't block the response. Will fix it.

shs96c commented on java/server/src/org/openqa/selenium/grid/sessionqueue/NewSessionQueuer.java:
```diff
@@ -0,0 +1,132 @@
+          span.addEvent(AttributeKey.EXCEPTION_EVENT.getKey(), attributeMap);
+          throw exception;
+        } else {
+          Capabilities caps = iterator.next();
```
We want to validate all the capabilities contained in the incoming request, not just the first.

shs96c commented on java/server/src/org/openqa/selenium/grid/sessionqueue/NewSessionQueuer.java:
```diff
@@ -0,0 +1,132 @@
+          throw exception;
+        } else {
+          Capabilities caps = iterator.next();
+          if (caps.getBrowserName().isEmpty()) {
```
This assumes that there won't be a slot stereotype that checks for something different, and we can't be sure of that. The best we can do for validation is a fast pass through the stereotypes the `Distributor` currently knows about, and if none of those could match the `Capabilities` then we _may_ have an error, but only if all the `Capabilities` in the new session request are unmatchable.

pujagani replied:
Alright. So for validation, we want to use the immutable model of hosts and see if any slots match any of the requested capabilities, if not we want to send back an error response?

shs96c commented on java/server/src/org/openqa/selenium/grid/sessionqueue/RemoveFromSessionQueue.java:
```diff
@@ -0,0 +1,45 @@
+        return response;
+      }
+
+      return new HttpResponse().setStatus(HTTP_NO_CONTENT);
```
Do we want to add the response to the span?

pujagani replied:
Will fix it. 👍

shs96c reviewed (changes requested):
The latches in the data transfer objects make me uneasy....

pujagani commented:
> The latches in the data transfer objects make me uneasy....

The latch is not being passed via event bus in any data transfer object. Instead, when a new session request is encountered and added to the queue, a localmap in GetNewSessionResponse.java stores the requestId, it's respective latch and empty response. The same latch now awaits the response. When the new session is created, it fires an event with the requestId, this is is picked up by the listener in GetNewSessionResponse. Now the response is set for that requestId, it's latch countdown is called and response returned to the caller. The requestId is now removed from the map.

diemol reviewed (approved):
Thanks, @pujagani!
Let's follow up in the next one by connecting the queue to the distributor and router.

pujagani commented:
Yes, will do that for the next one. Thanks a lot, @diemol !
